weave/flow/dataset.py

```diff
@@ -181,3 +199,178 @@ def select(self, indices: Iterable[int]) -> Self:
+                UserWarning,
+                stacklevel=2,
+            )
+
+        async def process_row(row: dict) -> dict:
+            try:
+                # Extract and pass specific parameters
+                kwargs = {}
+                for name in param_names:
+                    if name in row:
+                        kwargs[name] = row[name]
+                    else:
+                        # Skip this parameter or use a default if available
+                        if sig.parameters[name].default is not inspect.Parameter.empty:
+                            kwargs[name] = sig.parameters[name].default
+                        else:
+                            raise ValueError(
+                                f"Function expects parameter '{name}' but this column "
+                                f"is not in the dataset row: {short_str(row)}"
+                            )
+
+                result = func(**kwargs)
+
+                # If the user function is async, await it
+                if asyncio.iscoroutine(result):
+                    result = await result
+
+                # Handle the result based on its type
+                if not isinstance(result, dict):
+                    # Raise error for non-dictionary returns
+                    raise TypeError(
+                        f"Function must return a dictionary, but got {type(result).__name__} "
+                        f"from function '{func.__name__}'. Use a dictionary to specify "
+                        f"which columns to add or update."
+                    )
+            except Exception as e:
+                # Log the error and propagate it to stop the map operation
+                print(f"Error processing row with map function: {short_str(row)}")
+                traceback.print_exc()
+                raise e
+            else:
+                # Update the original row with the results
+                row_copy = row.copy()
+                row_copy.update(result)
+                return row_copy
+
+        async def main(progress: Progress) -> "Dataset":
+            async for _, processed_row in async_foreach(
+                self.rows,
+                process_row,
+                num_procs,
+                progress=progress,
+                progress_desc="Mapping dataset",
+            ):
+                processed_rows.append(processed_row)
+
+            return self.__class__(rows=processed_rows)
+
+        # Setup and run the progress bar and async processing
+        with Progress(
+            TextColumn("[progress.description]{task.description}"),
+            MofNCompleteColumn(),
+            TaskProgressColumn(),
+            BarColumn(),
+            IterationSpeedColumn(),
+            TimeElapsedColumn(),
+            console=console,  # Use the instantiated console
+            transient=True,  # Remove progress bar when done
+        ) as progress:
+            return asyncio.run(main(progress))
```

Comment:
Is this safe in all contexts?  e.g. does it work in a notebook, existing async event loop, etc?

Reply:
haven't tests =P